File: python/day10_part1_2021.py
with open('input.txt', 'r') as file:
    lines = file.readlines()

illegal_chars = {')': 3, ']': 57, '}': 1197, '>': 25137}
total_score = 0

for line in lines:
    stack = []
    for i, char in enumerate(line):
        if char in '([{<':
            stack.append(char)
        elif char in ')]}>':
            if not stack or ord(char) - ord(stack.pop()) not in [1, 2]:
                total_score += illegal_chars[char]
                break

print(total_score)

File: matlab/day10_part1_2021.m

function main()
    fid = fopen('input.txt', 'r');
    lines = textscan(fid, '%s', 'Delimiter', '\n');
    fclose(fid);
    lines = lines{1};

    illegal_chars = containers.Map({')', ']', '}', '>'}, {3, 57, 1197, 25137});
    total_score = 0;

    for i = 1:length(lines)
        line = lines{i};
        stack = {};
        for j = 1:length(line)
            char = line(j);
            if ismember(char, {'(', '[', '{', '<'})
                stack{end+1} = char;
            elseif ismember(char, {')', ']', '}', '>'})
                if isempty(stack) || ~((strcmp(char, ')') && strcmp(stack{end}, '(')) || ...
                                       (strcmp(char, ']') && strcmp(stack{end}, '[')) || ...
                                       (strcmp(char, '}') && strcmp(stack{end}, '{')) || ...
                                       (strcmp(char, '>') && strcmp(stack{end}, '<')))
                    total_score = total_score + illegal_chars(char);
                    break;
                else
                    stack(end) = [];
                end
            end
        end
    end

    fprintf('%d\n', total_score);
end
